### PR TITLE
Add info for creating SP initiated logout requests when name ID forma…

### PR DESCRIPTION
**Pull request by Yu-Chieh-Henry-Yang**
…t is not transient

### CODE REVIEW

**Comment by mattbaumann1** [comment]
That is good news.

File: app/controllers/devise/saml_sessions_controller.rb [hunk 1]
```diff
@@ -51,11 +53,31 @@ def relay_state
...
+    @name_identifier_value_for_sp_initiated_logout = current_user.send(Devise.saml_default_user_key)
+    @sessionindex_for_sp_initiated_logout = current_user.send(Devise.saml_session_index_key)
+  end
+
   # Override devise to send user to IdP logout for SLO
   def after_sign_out_path_for(_)
     idp_entity_id = get_idp_entity_id(params)
     request = OneLogin::RubySaml::Logoutrequest.new
-    request.create(saml_config(idp_entity_id))
+    saml_settings = saml_config(idp_entity_id)
```
[the remark this answers is not in the document]
Comment (Yu-Chieh-Henry-Yang) [reply app/controllers/devise/saml_sessions_controller.rb]: Sorry, I'm lost here. Can you please explain which value we are talking about here and when might it be mutated and causing an issue?

File: app/controllers/devise/saml_sessions_controller.rb [hunk 2]
```diff
@@ -51,11 +53,31 @@ def relay_state
     end
   end
 
+  # For non transient name ID, save info to identify user for logout purpose
+  # before that user's session got destroyed. These info are used in the
+  # `after_sign_out_path_for` method below.
+  def store_info_for_sp_initiated_logout
+    return if Devise.saml_config.name_identifier_format == 'urn:oasis:names:tc:SAML:2.0:nameid-format:transient'
+    @name_identifier_value_for_sp_initiated_logout = current_user.send(Devise.saml_default_user_key)
+    @sessionindex_for_sp_initiated_logout = current_user.send(Devise.saml_session_index_key)
```
[the remark this answers is not in the document]
Comment (Yu-Chieh-Henry-Yang) [reply app/controllers/devise/saml_sessions_controller.rb]: Yes, I have updated it

File: app/controllers/devise/saml_sessions_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamstegman) [reply app/controllers/devise/saml_sessions_controller.rb]: The [return value of `saml_config`](https://github.com/apokalipto/devise_saml_authenticatable/blob/f310efdf49f4010b7ce47eb08eb75f3dc016b3e5/lib/devise_saml_authenticatable/saml_config.rb#L8) can be the global config created in `devise.rb`. In this method, you assign a new value to that global config.

So I'm suggesting you `dup` this value so you're not mutating the global config from `devise.rb`.

```suggestion
    saml_settings = saml_config(idp_entity_id).dup
```

**Comment by mattbaumann1** [comment]
I have been testing this branch and there are still some issues with it or maybe these are old issues that are reappearing. See my comment on issue #117. I am also having some issues with the saml_session_index_key, I am not sure if this relates back to issue #22.

**Pull request merged**